woocommerce/woocommerce-gateway-stripe
# Enable saved payments from setup intents for payment methods that are converted to SEPA

# Changes proposed in this Pull Request:

Issue: Link to the GitHub issue this PR addresses (if appropriate).
Fixes #2010 
Fixes #2035 

Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.

Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.

This PR fixes the saved payment flow with setup intents for payment methods that are converted to SEPA Direct Debits. When saving one of these payment methods using a payment intent, we need to scour the payment method details within the charge on the intent to find the generated SEPA payment method ID that we would like to store and reuse. Setup intents do not have charges, but instead there is a secret field called `latest_attempt` that we can expand to find the corresponding generated payment method details. 

This PR also changes the saved payment flow in a more general manner. Previously, we created a WC Payment Token and stored it during the checkout: that is actually not necessary, as whenever we query for WC Payment Tokens, we additionally query Stripe to see if new payment methods have been added and create and store tokens for these new payment methods on the fly. To support this, I've removed the function call to create a token at checkout and now we simply clear the customer's cache, to ensure that we make a new request to query Stripe for payment methods, rather than using the saved transient.

This change also allowed me to significantly thin out and refactor the UPE payment method class. I've added unit test coverage for all the flows in this PR and update the existing tests to play along as well. Additionally, while #2010 only specifically mentions Bancontact, I have also set the iDEAL and Sofort payment methods to be reusable as well, as they should work the same way. If you think I might be acting too big for my boots, please let me know and I'll yield and switch this property for these other payment methods back to `false`.

# Testing instructions

Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?

1. Enable UPE and Bancontact within UPE.
2. Set up WooCommerce Subscriptions with a subscription product with a free trial.
3. Go to checkout with that product and use Bancontact to make the payment.
4. Authorize the transaction. 
5. Order should complete successfully and the Order Received page should show the correct payment method. 
6. Select the Subscription (either from the Order Received page or your Accounts page) and confirm that the Bancontact payment method is listed as the payment method for the subscription.
7. Ensure that this payment method is saved as a SEPA Direct Debit in your Account's list of saved payment methods.
8. Ensure that you can complete a checkout with this saved payment method.

Other things that are perhaps worth testing as well
 - Purchasing a regular subscription product (without a free-trial and with a payment intent) with Bancontact.
 - Purchasing a regular product and saving the payment method with Bancontact.
 - All of the above with iDEAL and SOFORT.
 - All existing saved payment method flows (sorry).
-------------------
